src/storage-types/src/sources.rs

```diff
@@ -1267,6 +1334,186 @@ impl Schema<SourceData> for RelationDesc {
     }
 }
 
+/// Describes how subsource references should be organized in a multi-level
+/// hierarchy.
+///
+/// For both PostgreSQL and MySQL sources, these levels of reference are
+/// intrinsic to the items which we're referencing. If there are other naming
+/// schemas for other types of sources we discover, we might need to revisit
+/// this.
+pub trait SubsourceCatalogReference {
+    /// The "second" level of namespacing for the reference.
+    fn schema_name(&self) -> &str;
+    /// The lowest level of namespacing for the reference.
+    fn item_name(&self) -> &str;
+}
+
+impl SubsourceCatalogReference for mz_mysql_util::MySqlTableDesc {
+    fn schema_name(&self) -> &str {
+        &self.schema_name
+    }
+
+    fn item_name(&self) -> &str {
+        &self.name
+    }
+}
+
+impl SubsourceCatalogReference for mz_postgres_util::desc::PostgresTableDesc {
+    fn schema_name(&self) -> &str {
+        &self.namespace
+    }
+
+    fn item_name(&self) -> &str {
+        &self.name
+    }
+}
+
+// This implementation provides a means of converting arbitrary objects into a
+// `SubsourceCatalogReference`, e.g. load generator view names.
+impl<'a> SubsourceCatalogReference for (&'a str, &'a str) {
+    fn schema_name(&self) -> &str {
+        self.0
+    }
+
+    fn item_name(&self) -> &str {
+        self.1
+    }
+}
+
+/// Stores and resolves references to a `&[T: SubsourceCatalogTable]`.
+///
+/// This is meant to provide an API to quickly look up a source's subsources.
+///
+/// For sources that do not provide any subsources, use the `Default`
+/// implementation, which is empty and will not be able to resolve any
+/// references.
+#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, Arbitrary)]
+pub struct SubsourceResolver {
+    inner: BTreeMap<String, BTreeMap<String, BTreeMap<String, usize>>>,
+}
+
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum SubsourceResolutionError {
+    #[error("reference to {name} not found in source")]
+    DoesNotExist { name: String },
+    #[error(
+        "reference to {name} is ambiguous, consider specifying an additional \
+    layer of qualification"
+    )]
+    Ambiguous { name: String },
+}
+
+impl<'a> SubsourceResolver {
+    pub fn new<T: SubsourceCatalogReference>(
+        database: &str,
+        referenceable_items: &'a [T],
+    ) -> SubsourceResolver {
+        // An index from table name -> schema name -> database name -> PostgresTableDesc
+        let mut inner = BTreeMap::new();
+
+        // Here and below, we want to treat these as if they came from an
+        // `UnresolvedItemName` (because the actual references in subsources
+        // will have), so filter them through the `Ident` API. This is fine to
+        // do because any string here must have already proven it's capable of
+        // being taken to an `Ident` in planning.
```

Comment:
```suggestion
        // do because any string here must have already proven it's capable of
        // being taken to an `Ident` in planning. Note that `to_string()` uses `Ident`'s `Display` impl,
        // which properly escapes the string.
```

Reply:
Realized there was a better way of doing this with a bit more elbow grease––just needed to actually define an `Arbitrary` impl that works on `Ident`. Thanks for the nudge here.